src/backend/distributed/planner/relation_restriction_equivalence.c

```diff
@@ -626,13 +628,27 @@ EquivalenceListContainsRelationsEquality(List *attributeEquivalenceList,
 		RelationRestriction *relationRestriction =
 			(RelationRestriction *) lfirst(relationRestrictionCell);
 		int rteIdentity = GetRTEIdentity(relationRestriction->rte);
+		int colocationId = TableColocationId(relationRestriction->relationId);
```

Comment:
I think this works too, but as noted below, some test outputs sounds less accurate now.
And, also this place seems too deep inside the planning to make the colocation checks to me.

One suggestion could be to make the colocation checks more explicit  on `SafeToPushdownUnionSubquery()` and `RestrictionEquivalenceForPartitionKeys`. To do  that, we  could consider  refactoring `ErrorIfUnsupportedShardDistribution()` function to  defer error, and on these functions return false if the relations are not colocated. 

To improve the  error messages ,  we can use the same  check on `DeferErrorIfUnsupportedSubqueryPushdown()`, before getting into the other checks.

I can see that we'd need  to enforce the same check in multiple places, but still wanted to push this comment in case you can propose  something simpler.

Otherwise, we can merge  this  patch

Reply:
I agree, would make sense to have a separate check.